### PR TITLE
Suggest move for closures and async blocks in more cases.

### LINKED ISSUE #66107: async-block diagnostics do not suggest async move when it may be needed
[(playground)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3d75e33c51d66cd9dd7a95d71648a5cd)

```rust
//* f hits lifetime errors
fn f(a: &usize) -> (impl '_ + Future<Output = ()>, &usize) {
    (async { *a; }, a)
}
// */

// g compiles as expected
fn g(a: &usize) -> (impl '_ + Future<Output = ()>, &usize) {
    async fn h(a: &usize) { *a; }
    (h(a), a)
}
```

```
error[E0515]: cannot return value referencing function parameter `a`
 --> src/main.rs:5:5
  |
5 |     (async { *a; }, a)
  |     ^^^^^^^^^^-^^^^^^^
  |     |         |
  |     |         `a` is borrowed here
  |     returns a value referencing data owned by the current function
```

This issue has been assigned to @gizmondo via [this comment](https://github.com/rust-lang/rust/issues/66107#issuecomment-610613109).

### PR DESCRIPTION
**Pull request by gizmondo**
Fixes #66107, also improves #67577
Related PR https://github.com/rust-lang/rust/pull/65166

### CODE REVIEW

File: src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -1225,8 +1204,9 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         );
 
         let msg = match category {
-            ConstraintCategory::Return => "closure is returned here".to_string(),
-            ConstraintCategory::OpaqueType => "generator is returned here".to_string(),
+            ConstraintCategory::Return | ConstraintCategory::OpaqueType => {
```
Comment (gizmondo) [inline src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs:1207]: I feel the existing behavior is wrong, users of async should not see `generator` in error messages.
But not sure what wording should be used instead.

File: src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs [hunk 2]
```diff
@@ -760,47 +760,26 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             (
                 Some(ref name),
                 BorrowExplanation::MustBeValidFor {
-                    category: category @ ConstraintCategory::Return,
+                    category:
+                        category
+                        @
+                        (ConstraintCategory::Return
+                        | ConstraintCategory::CallArgument
+                        | ConstraintCategory::OpaqueType),
```
Comment (Centril) [inline src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs:768]: Highlighting this for a rustfmt issue I'm filing (no need to do anything about it in this PR).

File: src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs (hunk 1 above)
Comment (estebank) [reply src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs:1207]: I think that we'll 1) need to be able to differentiate between closures and generators here in the first place and 2) in the `OpaqueType` it should be `"async closure is returned here"`, right?

File: src/test/ui/async-await/async-borrowck-escaping-block-error.stderr [hunk 3]
```diff
@@ -7,16 +7,35 @@ LL |     Box::new(async { x } )
...
+   |
+note: closure is returned here
+  --> $DIR/async-borrowck-escaping-block-error.rs:11:5
+   |
+LL |     async { *x }
+   |     ^^^^^^^^^^^^
+help: to force the async block to take ownership of `x` (and any other referenced variables), use the `move` keyword
+   |
+LL |     async move { *x }
+   |           ^^^^^^^^^^^
```
Comment (estebank) [inline src/test/ui/async-await/async-borrowck-escaping-block-error.stderr:37]: Actually, I think the output here should talk about closures, it should talk about async blocks instead (same in the previously existing test):

```

error[E0373]: async block may outlive the current function, but it borrows `x`, which is owned by the current function
  --> $DIR/async-borrowck-escaping-block-error.rs:11:11
   |
LL |     async { *x }
   |           ^^^-^^
   |           |  |
   |           |  `x` is borrowed here
   |           may outlive borrowed value `x`
   |
note: async block is returned here
  --> $DIR/async-borrowck-escaping-block-error.rs:11:5
   |
LL |     async { *x }
   |     ^^^^^^^^^^^^
help: to force the async block to take ownership of `x` (and any other referenced variables), use the `move` keyword
   |
LL |     async move { *x }
   |           ^^^^^^^^^^^
```

Comment (gizmondo) [reply src/test/ui/async-await/async-borrowck-escaping-block-error.stderr]: Updated, thanks.

File: src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/diagnostics/conflict_errors.rs:1209]: Agreed that saying generators in an error message is basically never right, regardless.

**Pull request merged**